removed preview from ManagedServiceAccount

ngraham20 opened the pull request:
Additionally, managedserviceaccount-preview will function in place of default during Reconcile loop

ngraham20 commented:
unit tests have not yet been updated

ngraham20 commented on api/v1/multiclusterengine_methods.go:
```diff
@@ -19,19 +19,20 @@ limitations under the License.
-	HypershiftLocalHosting = "hypershift-local-hosting"
-	LocalCluster           = "local-cluster"
+	ManagedServiceAccount        = "managedserviceaccount"
+	ManagedServiceAccountPreview = "managedserviceaccount-preview"
```
This is the actual change, the rest is just tabbing

JakobGray commented on controllers/backplaneconfig_controller.go:
```diff
@@ -517,7 +517,8 @@ func (r *MultiClusterEngineReconciler) ensureToggleableComponents(ctx context.Co
 
-	if backplaneConfig.Enabled(backplanev1.ManagedServiceAccount) {
+	if backplaneConfig.Enabled(backplanev1.ManagedServiceAccount) ||
+		backplaneConfig.Enabled(backplanev1.ManagedServiceAccountPreview) {
```
This is still taking into account the preview's configuration, which doesn't align with the criteria
> managedserviceaccount-preview component still recognized, but ignored.

JakobGray reviewed (changes requested):
When a new component is added to backplane-operator, it also needs to be added to mch-operator to recognize it https://github.com/stolostron/multiclusterhub-operator/blob/main/api/v1/multiclusterhub_methods.go#L25

ngraham20 commented:
/retest

JakobGray commented on controllers/backplaneconfig_controller.go:
```diff
@@ -1038,6 +1038,13 @@ func (r *MultiClusterEngineReconciler) setDefaults(ctx context.Context, m *backp
 
+	// managedserviceaccount preview component upgraded in 2.9.0
+	if m.Prune(backplanev1.ManagedServiceAccountPreview) {
+		// if the preview was pruned, enable the non-preview version instead
```
This sounds like it will always enable managedSA on an upgrade since msa-preview will always be there and need pruning. It should only enable managedSA if msa-preview was also enabled.

JakobGray reviewed (commented):
I recommend writing a unit test passing in an MCE with the old msa-preview and verifying that it
1) removes the preview component
2) adds the non-preview
3) enables the non-preview by default if preview was enabled

ngraham20 commented:
I found what's causing the tests to fail:

```go
func SetDefaultComponents(m *backplanev1.MultiClusterEngine) bool {
	updated := false
	for _, c := range onComponents {
		if !m.ComponentPresent(c) {
			m.Enable(c)
			updated = true
		}
	}
	for _, c := range offComponents {
		if !m.ComponentPresent(c) { // THIS IS THE ISSUE
			m.Disable(c)
			updated = true
		}
	}
	return updated
}
```
the line `m.Disable(c)` will create any component in the offComponent list that it doesn't already see, but the `Prune()` statement which occurs after this function call will delete it again. Endless cycle.

Do you think we can remove `ManagedServiceAccountPreview` from the offComponents list to avoid this?

ngraham20 commented:
/retest

dislbenn commented:
/lgtm

dislbenn commented:
/hold

dislbenn commented:
Related issue: https://issues.redhat.com/browse/ACM-7008

dislbenn commented:
/unhold
